### PR TITLE
sql: Support `IS BOOLEAN` operators

### PR DESCRIPTION
**Pull request by cjubb39**
### Motivation
Fixes https://github.com/MaterializeInc/materialize/issues/1052

### Description
Add support for `IS {TRUE, FALSE, UNKNOWN}` and `IS NOT {TRUE, FALSE, UNKNOWN}`.  This is done by expanding the `IsNull` expression that currently exists to support the other constructs.  Additionally add new functions to compute the value of these UnaryFunctions.

### Tips for reviewer
First commit has parser changes.  Second has plan changes.

### Checklist

- [x] This PR has adequate test coverage.
- [ ] This PR adds a release note for any user-facing behavior changes.

### CODE REVIEW

File: src/expr/src/scalar/func/impls/datum.rs [hunk 1]
```diff
@@ -21,6 +21,36 @@ sqlfunc!(
...
+    }
+);
+
+sqlfunc!(
+    #[sqlname = "isunknown"]
+    #[propagates_nulls = false]
+    #[introduces_nulls = false]
+    #[preserves_uniqueness = false]
+    fn is_unknown(a: Datum<'_>) -> Result<Option<bool>, EvalError> {
+        Ok(Some(a == Datum::Null))
```
Comment (cjubb39) [inline src/expr/src/scalar/func/impls/datum.rs:50]: I think this is correct given the docs here: https://www.postgresql.org/docs/9.1/functions-comparison.html.

Though I'm not sure how to do / if we need to do the "except that the input expression must be of Boolean type"?

File: src/sql/src/plan/query.rs [hunk 2]
```diff
@@ -3120,27 +3124,36 @@ pub fn resolve_func(
...
     // unconstrained parameters identically. Providing a type hint of string
     // means we wind up supporting both.
     let expr = plan_expr(ecx, inner)?.type_as_any(ecx)?;
-    let mut expr = HirScalarExpr::CallUnary {
-        func: UnaryFunc::IsNull(expr_func::IsNull),
+    let func = match construct {
+        IsExprConstruct::NULL => UnaryFunc::IsNull(expr_func::IsNull),
+        IsExprConstruct::TRUE => UnaryFunc::IsTrue(expr_func::IsTrue),
+        IsExprConstruct::FALSE => UnaryFunc::IsFalse(expr_func::IsFalse),
+        IsExprConstruct::UNKNOWN => UnaryFunc::IsUnknown(expr_func::IsUnknown),
```
Comment (sploiselle) [inline src/sql/src/plan/query.rs:3142]: Here you'll want to use the `ecx` in this function to check the scalartype of `expr`; should look something like `matches!(ecx.scalar_type(&expr), ScalarType::Bool)`, and then you'll need to error if not. lmk if you want any help on the front end of that work.

File: src/expr/src/scalar/func/impls/datum.rs (hunk 1 above)
Comment (sploiselle) [reply src/expr/src/scalar/func/impls/datum.rs:50]: Yeah, it's a little tricky with these because they're "expressions" and not part of our normal function selection framework.

The gist is that you'll have to make sure that any expression that gets passed to this function has the right `ScalarType`. In most cases that's handled for you, but here we'll need to handroll a type check [here](https://github.com/MaterializeInc/materialize/pull/8455/files#diff-28dcff29ce185a34dc7bb2444006600d245dd969d5bcd3a4e9bef65a36b5efc1R3142).

File: src/sql-parser/src/ast/defs/expr.rs [hunk 3]
```diff
@@ -59,7 +59,11 @@ pub enum Expr<T: AstInfo> {
         right: Box<Expr<T>>,
     },
     /// `IS NULL` expression
```
Comment (sploiselle) [inline src/sql-parser/src/ast/defs/expr.rs:61]: doc comment needs to be adjusted for expansion in scope

File: src/sql-parser/src/ast/defs/expr.rs [hunk 4]
```diff
@@ -735,3 +743,22 @@ impl<T: AstInfo> AstDisplay for FunctionArgs<T> {
     }
 }
 impl_display_t!(FunctionArgs);
+
+#[derive(Debug, Clone, Copy, PartialEq, Eq, Hash)]
+pub enum IsExprConstruct {
+    NULL,
+    TRUE,
+    FALSE,
+    UNKNOWN,
```
Comment (benesch) [inline src/sql-parser/src/ast/defs/expr.rs:752]: Rust style is to use UpperCamelCase for enum variants rather than SNAKE_CASE.

File: src/expr/src/scalar/func/impls/datum.rs [hunk 5]
```diff
@@ -21,6 +21,36 @@ sqlfunc!(
...
+    }
+);
+
```
Comment (benesch) [inline src/expr/src/scalar/func/impls/datum.rs:53]: So actually no need to introduce new UnaryFunc variants in this case! Instead you can plan `IS TRUE` as `BinaryFunc::Eq(<expr>, Datum::True)`, `IS FALSE` as `BinaryFunc::Eq(<expr>, Datum::False)`, and `IS UNKNOWN` as `UnaryFunc::IsNull(<expr>)`.

File: src/sql/src/plan/query.rs (hunk 2 above)
Comment (benesch) [reply src/sql/src/plan/query.rs:3142]: I think actually push the call to `type_as_any` into the `match`, and then change it to `type_as(..., ScalarType::Boolean)` for `IsExprConstruct::{True, False, Unknown}`, but leave it as `type_as_any` for `NULL` (to capture the difference between `IS NULL` and `IS UNKNOWN`).

**Review by benesch — COMMENTED** [review]
🎉

File: src/sql-parser/src/ast/defs/expr.rs (hunk 4 above)
Comment (cjubb39) [reply src/sql-parser/src/ast/defs/expr.rs:752]: ahhhh it's been too long since I've written rust -- thanks

File: src/expr/src/scalar/func/impls/datum.rs (hunk 5 above)
Comment (cjubb39) [reply src/expr/src/scalar/func/impls/datum.rs:53]: I took a stab at this route and unfortunately `BinaryFunc::Eq` propagates NULLs which is not the behavior we want here.  e.g. `NULL IS TRUE` should return `false` not `NULL`.  It seemed worse to introduce a generic `EqWithoutPropagatingNull` function than to have a couple special functions.  What do you think?

I did go ahead and get rid of the special `IsUnknown`

Comment (benesch) [reply src/expr/src/scalar/func/impls/datum.rs]: Aw shoot, sorry, I forgot #7597 hasn't been merged. Can you instead drop a TODO that `IsTrue` and `IsFalse` can be replaced with `BinaryFunc::eq` once #7581 is fixed?

File: src/sql/src/plan/query.rs [hunk 6]
```diff
@@ -3120,27 +3124,49 @@ pub fn resolve_func(
...
+        let expr_coerced= typeconv::plan_coerce(ecx, planned_expr, &ScalarType::Bool)?;
+        let expr_ty = ecx.scalar_type(&expr_coerced);
+        // PostgreSQL allows booleans themselves or string representations of booleans only.
+        // Even though we can coerce an integer to a boolean, this should not succeed.
+        if expr_ty == ScalarType::Bool || expr_ty == ScalarType::String {
+            expr_coerced
+        } else {
+            let not_msg = if not { "NOT " } else { "" };
+            bail!("IS{} {} requires boolean expression", not_msg, construct)
+        }
```
Comment (benesch) [inline src/sql/src/plan/query.rs:3144]: This looks funny to me! Did `planned_expr.type_as(ecx, ScalarType::Bool)` not do the right thing?

In particular, Postgres should allow any untyped string literal to coerce to a bool here, but not literals that have been explicitly typed as strings:

```
benesch=# select 'true'::text is true;
ERROR:  argument of IS TRUE must be type boolean, not type text
LINE 1: select 'true'::text is true;
               ^
benesch=# select 'true' is true;
 ?column? 
----------
 t
(1 row)

```

Comment (sploiselle) [reply src/sql/src/plan/query.rs:3144]: Ah, part of this is my fault. Chae and I were winding down for the day and I brought up `plan_coerce`, not thinking we could similarly achieve the same thing with one condition executing `type_as` and one executing `type_as_any`.

File: src/sql-parser/src/ast/defs/expr.rs [hunk 7]
```diff
@@ -735,3 +743,42 @@ impl<T: AstInfo> AstDisplay for FunctionArgs<T> {
...
+impl fmt::Display for IsExprConstruct {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        match self {
+            IsExprConstruct::Null => f.write_str("NULL"),
+            IsExprConstruct::True => f.write_str("TRUE"),
+            IsExprConstruct::False => f.write_str("FALSE"),
+            IsExprConstruct::Unknown => f.write_str("UNKNOWN"),
+        }
+    }
+}
```
Comment (benesch) [inline src/sql-parser/src/ast/defs/expr.rs:773]: Oh, this should be auto-derived using the `impl_display!` macro, which just delegates to the `AstDisplay` impl.

File: src/sql/src/plan/query.rs [hunk 8]
```diff
@@ -3120,27 +3124,49 @@ pub fn resolve_func(
...
+    let planned_expr = plan_expr(ecx, inner)?;
+    let expr = if construct.requires_boolean_expr() {
+        let expr_coerced= typeconv::plan_coerce(ecx, planned_expr, &ScalarType::Bool)?;
+        let expr_ty = ecx.scalar_type(&expr_coerced);
+        // PostgreSQL allows booleans themselves or string representations of booleans only.
+        // Even though we can coerce an integer to a boolean, this should not succeed.
+        if expr_ty == ScalarType::Bool || expr_ty == ScalarType::String {
+            expr_coerced
+        } else {
+            let not_msg = if not { "NOT " } else { "" };
```
Comment (ruchirK) [inline src/sql/src/plan/query.rs:3142]: minor: I think this should be `" NOT"` instead of `"NOT "` based on the error string below. The way its currently written seems like it would print something like:

```
"ISNOT TRUE requires boolean expression"
```

File: src/sql/src/plan/query.rs (hunk 6 above)
Comment (cjubb39) [reply src/sql/src/plan/query.rs]: ahh you're right, that's the correct way to do it.

I think the implementation of that made me think it wouldn't quite work:  the difference between `expr` and `expr_ty` didn't make sense to me.  I think I was confused by what was actually happening in `plan_coerce`

Comment (sploiselle) [reply src/sql/src/plan/query.rs]: Yeah, `CoercibleScalarExpr` takes a little bit of exposure to make sense of. It's all in service of letting us remain agnostic about the type of things that come in as untyped string literals.

File: src/sql-parser/src/parser.rs [hunk 9]
```diff
@@ -947,27 +947,43 @@ impl<'a> Parser<'a> {
         } else if let Token::Keyword(kw) = tok {
             match kw {
                 IS => {
+                    let negated = self.parse_keyword(NOT);
                     if self.parse_keyword(NULL) {
```
Comment (sploiselle) [inline src/sql-parser/src/parser.rs:951]: This can be reduced a good bit by using `parse_one_of_keywords` and returning the expected `construct`. You can find some inspiration for this approach in [`parse_create_role`](https://github.com/MaterializeInc/materialize/blob/main/src/sql-parser/src/parser.rs#L2118-L2124).

File: test/sqllogictest/comparison.slt [hunk 10]
```diff
@@ -28,3 +28,47 @@ query B
...
+
+query error invalid input syntax for type boolean
+select '1 = 1' is true
+
+query B
+select '1 = 1' is null
+----
+false
+
+query error
```
Comment (sploiselle) [inline test/sqllogictest/comparison.slt:73]: we should propagate the error string we're matching on to this test as the others

File: test/sqllogictest/comparison.slt [hunk 11]
```diff
@@ -28,3 +28,47 @@ query B
 select null isnull
 ----
 true
+
```
Comment (sploiselle) [inline test/sqllogictest/comparison.slt:31]: Let's add a test case that shows that we're handling the difference between `is null` and `is unknown`

**Review by sploiselle — COMMENTED** [review]
Looks just about done!

For commits, I'd suggest taking a look at `doc/developer/guide-changes.md`; in short, just want to make sure your commit messages ultimately express which part of the code you're touching and how.

**Comment by sploiselle** [comment]
LGTM modulo info about the commit message. Will also need to write a release note (can be a separate PR), and either draft some docs or file an issue for the docs to get written.

**Comment by benesch** [comment]
(LGTM too, but deferring approval to Sean!)
